url901cc5721b_shidasan_KonohaCoffee_tgz-pJ8-org_KonohaScript_KonohaJ8/src/org/KonohaScript/CodeGen/SimpleVMCodeGen.java

```diff
@@ -0,0 +1,423 @@
+		String Right = pop();
+		String Left = pop();
+		push(Left + " && " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterOr(OrNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitOr(OrNode Node) {
+		String Right = pop();
+		String Left = pop();
+		push(Left + " || " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterAssign(AssignNode Node) {
+		AddLocalVarIfNotDefined(Node.TermToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitAssign(AssignNode Node) {
+		String Right = pop();
+		push(Node.TermToken.ParsedText + " = " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterLet(LetNode Node) {
+		AddLocalVarIfNotDefined(Node.TermToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitLet(LetNode Node) {
+		String Block = pop();
+		String Right = pop();
+		push(Node.TermToken.ParsedText + " = " + Right + Block);
+		return true;
+
+	}
+
+	@Override
+	public void EnterBlock(BlockNode Node) {
+		this.CurrentProgramSize = this.Program.size();
+	}
+
+	@Override
+	public boolean ExitBlock(BlockNode Node) {
+		String Exprs = "";
+		int Size = this.Program.size() - this.CurrentProgramSize;
+		for (int i = 0; i < Size; i = i + 1) {
+			String Expr = pop();
+			Exprs = Expr + ";" + Exprs;
+		}
+		push("{" + Exprs + "}");
+		return true;
+
+	}
+
+	@Override
+	public void EnterIf(IfNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitIf(IfNode Node) {
+		String ElseBlock = pop();
+		String ThenBlock = pop();
+		String CondExpr = pop();
+		push("if (" + CondExpr + ") " + ThenBlock + " else " + ElseBlock);
+		return true;
+
+	}
+
+	@Override
+	public void EnterSwitch(SwitchNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitSwitch(SwitchNode Node) {
+		int Size = Node.Labels.size();
+		String Exprs = "";
+		for (int i = 0; i < Size; i = i + 1) {
+			String Label = Node.Labels.get(Size - i);
+			String Block = pop();
+			Exprs = "case " + Label + ":" + Block + Exprs;
+		}
+		String CondExpr = pop();
+		push("switch (" + CondExpr + ") {" + Exprs + "}");
+		return true;
+	}
+
+	@Override
+	public void EnterLoop(LoopNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitLoop(LoopNode Node) {
+		String LoopBody = pop();
+		String IterExpr = pop();
+		String CondExpr = pop();
+		push("while (" + CondExpr + ") {" + LoopBody + IterExpr + "}");
+		return true;
+
+	}
+
+	@Override
+	public void EnterReturn(ReturnNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitReturn(ReturnNode Node) {
+		String Expr = pop();
+		push("return " + Expr);
+		return false;
+
+	}
+
+	@Override
+	public void EnterLabel(LabelNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitLabel(LabelNode Node) {
+		String Label = Node.Label;
+		if (Label.compareTo("continue") == 0) {
+			push("");
+		} else if (Label.compareTo("continue") == 0) {
+			push("");
+		} else {
+			push(Label + ":");
+		}
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 336:344, 355:363.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.